### PR TITLE
[mqtt] Add IMap persistence for source

### PR DESCRIPTION
**Pull request by gurbuzali**
#### Which issue this PR fixes
  - fixes https://github.com/hazelcast/hazelcast-jet-contrib/issues/104
  - fixes https://github.com/hazelcast/hazelcast-jet-contrib/issues/102

#### Checklist

- [ ] Signed the Hazelcast CLA
- [ ] No checkstyle issues (`./gradlew check`)
- [ ] No tests failures (`./gradlew test`)
- [ ] Documentation which complies with README template (see `templates/README.template.md`).
- [ ] Update `List of modules` section on the root README with your module.

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/388/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/388/jet-contrib-pr-builder-388.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/389/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/389/jet-contrib-pr-builder-389.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHook.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                    return MQTT_PERSISTABLE;
+                }
+
+                @Override
+                public void write(ObjectDataOutput out, MqttPersistable persistable) throws IOException {
+                    try {
+                        byte[] header = new byte[persistable.getHeaderLength()];
+                        int headerOffset = persistable.getHeaderOffset();
+                        byte[] payload = new byte[persistable.getPayloadLength()];
+                        int payloadOffset = persistable.getPayloadOffset();
```
Comment (viliam-durina) [inline mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHook.java:58]: The payload is apparently optional: https://github.com/eclipse/paho.mqtt.java/blob/6f35dcb785597a6fd49091efe2dba47513939420/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/persist/MqttDefaultFilePersistence.java#L181-L182

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHook.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                        System.arraycopy(persistable.getHeaderBytes(), headerOffset, header, 0, header.length);
+                        System.arraycopy(persistable.getPayloadBytes(), payloadOffset, payload, 0, payload.length);
+                        out.writeByteArray(header);
```
Comment (viliam-durina) [inline mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHook.java:61]: Can't we use and avoid copying the array?
```
out.write(message.getHeaderBytes(), message.getHeaderOffset(), message.getHeaderLength());
```

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSourceBuilder.java [hunk 3]
```diff
@@ -118,6 +119,29 @@
...
+     * <p>
+     * By default the client and broker will not remember the state. If
+     * this parameter is set, the source ignores the value of
+     * {@link MqttConnectOptions#isCleanSession()} specified via
+     * {@link #connectOptionsFn(SupplierEx)}.
+     * <p>
+     * See {@link MqttConnectOptions#setCleanSession(boolean)}.
+     */
+    @Nonnull
+    public MqttSourceBuilder<T> persistSession(@Nonnull String mapName) {
```
Comment (viliam-durina) [inline mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSourceBuilder.java:140]: Couldn't we generate the map name automatically? The user needs to destroy the map after the job is done, but it's not easy to do.

**Review by viliam-durina — CHANGES_REQUESTED** [review]
We need a fault tolerance stress tests, [such as these](https://github.com/hazelcast/hazelcast-jet/blob/c1fc292d5226a41347ab86bf7b9b63f09dcde25f/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/WriteFilePTest.java#L359-L374). For both source and sink. Otherwise we don't know our implementation is correct.

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/390/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/390/jet-contrib-pr-builder-390.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

**Review by viliam-durina — COMMENTED** [review]
We should also explain what's needed for ex-once: to set the QoS, to set the map name.

**Pull request closed without merging**